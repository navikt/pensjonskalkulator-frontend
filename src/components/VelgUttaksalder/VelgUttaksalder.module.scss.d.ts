import globalClassNames from '../../style.d'
declare const classNames: typeof globalClassNames & {
  readonly wrapper: 'wrapper'
  readonly pin: 'pin'
  readonly chipsWrapper: 'chipsWrapper'
  readonly chipsWrapper__hasGap: 'chipsWrapper__hasGap'
  readonly visFlereAldere: 'visFlereAldere'
}
export = classNames
