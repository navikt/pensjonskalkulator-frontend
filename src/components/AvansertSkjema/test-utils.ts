export function getPreviousMonth(): string {
  const today = new Date()
  const currentMonth = parseInt(
    today.toLocaleDateString('en-CA').slice(5, 7),
    10
  ) // Extracts the month from the date string
  const previousMonth = currentMonth - 1
  return previousMonth.toString()
}
