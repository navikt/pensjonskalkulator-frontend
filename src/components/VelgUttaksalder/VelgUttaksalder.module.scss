@import '../../scss/variables';

.wrapper {
  display: flex;
  flex-direction: column;
  gap: var(--a-spacing-3);
}

.pin {
  position: absolute;
  margin-top: -2rem;
}

.chipsWrapper {
  /* double-specificity n√∏dvendig pga Aksel overskrivning */
  &.chipsWrapper__hasGap {
    gap: var(--a-spacing-3);

    &::after {
      content: '';
      flex: auto;
    }
  }

  :global {
    .navds-chips--icon-left {
      padding-left: var(--a-spacing-3);
    }

    .navds-chips__toggle-icon {
      display: none;
    }
  }
}

.visFlereAldere {
  margin-right: auto;
  margin-bottom: var(--a-spacing-3);
}
