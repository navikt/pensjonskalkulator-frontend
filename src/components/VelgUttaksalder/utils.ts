export const formatUttaksalder = (
  { aar, maaned }: Uttaksalder,
  options: { compact: boolean } = { compact: false }
): string => {
  return maaned !== 0
    ? `${aar} år og ${maaned} ${options.compact ? 'md.' : 'måneder'}`
    : `${aar} år`
}

export const getFormaterteAldere = (
  start: Uttaksalder,
  end: Uttaksalder = { aar: 75, maaned: 0 }
): string[] => {
  if (end.aar < start.aar) {
    return []
  }

  const aldere: string[] = [formatUttaksalder(start, { compact: true })]

  for (let i = start.aar + 1; i <= end.aar; i++) {
    aldere.push(formatUttaksalder({ aar: i, maaned: 0 }))
  }

  return aldere
}
