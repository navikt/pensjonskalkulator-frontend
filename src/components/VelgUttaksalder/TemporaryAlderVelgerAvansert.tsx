/* c8 ignore start */
import React from 'react'
import { useIntl } from 'react-intl'

import { Select } from '@navikt/ds-react'

import { useTidligsteUttaksalderQuery } from '@/state/api/apiSlice'
import { useAppSelector } from '@/state/hooks'
import {
  selectAfp,
  selectSamboer,
  selectSivilstand,
  selectAarligInntektFoerUttak,
} from '@/state/userInput/selectors'
import { formatUttaksalder, unformatUttaksalder } from '@/utils/alder'

import { DEFAULT_TIDLIGST_UTTAKSALDER, getFormaterteAldere } from './utils'

interface Props {
  name: string
  label: string
  description?: string
  defaultValue?: Alder | null
  value?: Alder | null // controlled component
  hasValidationError?: boolean
  onChange?: (alder: Alder | undefined) => void
}

export const TemporaryAlderVelgerAvansert: React.FC<Props> = ({
  name,
  label,
  description,
  defaultValue,
  value,
  hasValidationError,
  onChange,
}) => {
  const intl = useIntl()
  const afp = useAppSelector(selectAfp)
  const harSamboer = useAppSelector(selectSamboer)
  const sivilstand = useAppSelector(selectSivilstand)
  const aarligInntektFoerUttak = useAppSelector(selectAarligInntektFoerUttak)

  const [tidligsteUttaksalderRequestBody, setTidligsteUttaksalderRequestBody] =
    React.useState<TidligsteUttaksalderRequestBody | undefined>(undefined)
  const [localValue, setLocalValue] = React.useState<string>(
    defaultValue ? formatUttaksalder(intl, defaultValue, { compact: true }) : ''
  )

  React.useEffect(() => {
    setTidligsteUttaksalderRequestBody({
      sivilstand: sivilstand,
      harEps: harSamboer !== null ? harSamboer : undefined,
      sisteInntekt: aarligInntektFoerUttak ?? 0,
      simuleringstype:
        afp === 'ja_privat' ? 'ALDERSPENSJON_MED_AFP_PRIVAT' : 'ALDERSPENSJON',
    })
  }, [afp, sivilstand, aarligInntektFoerUttak, harSamboer])

  React.useEffect(() => {
    if (value !== undefined) {
      setLocalValue(
        value !== null ? formatUttaksalder(intl, value, { compact: true }) : ''
      )
    }
  }, [value])

  // Hent tidligst mulig uttaksalder
  const {
    data: tidligstMuligUttak,
    isLoading,
    isSuccess,
  } = useTidligsteUttaksalderQuery(tidligsteUttaksalderRequestBody, {
    skip: !tidligsteUttaksalderRequestBody,
  })

  const [showValidationError, setShowValidationError] =
    React.useState<boolean>(true)

  React.useEffect(() => {
    setShowValidationError(true)
  }, [hasValidationError])

  const onLocalChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    ;(prevShowValidationError: boolean) =>
      setShowValidationError(!prevShowValidationError)
    if (value === undefined) {
      setLocalValue(e.target.value)
    }
    if (onChange) {
      onChange(e.target.value ? unformatUttaksalder(e.target.value) : undefined)
    }
  }

  const formaterteAldere = React.useMemo(
    () =>
      getFormaterteAldere(
        intl,
        isSuccess ? tidligstMuligUttak : { ...DEFAULT_TIDLIGST_UTTAKSALDER }
      ),
    [tidligstMuligUttak]
  )

  if (isLoading) {
    return 'LOADING'
  }

  return (
    <div>
      {
        // TODO under avklaring - hvordan skal feil ved henting av tidligste uttaksalder vises?
        /* {isTidligstMuligUttaksalderError && (
          <BodyLong size="medium" className={`${styles.ingress}`}>
            <FormattedMessage
              id="tidligsteuttaksalder.error"
              values={{
                ...getFormatMessageValues(intl),
              }}
            />
          </BodyLong>
        )} */
      }
      <Select
        data-testid={`temporaryAlderVelger-${name}`}
        form="avansert-beregning"
        name={name}
        label={label}
        description={
          description !== undefined
            ? description
            : isSuccess && tidligstMuligUttak
              ? `Du kan tidligst ta ut 100 % alderspensjon når du er ${formatUttaksalder(
                  intl,
                  tidligstMuligUttak
                )}. Vil du ta ut pensjon tidligere, må du velge lavere uttaksgrad.`
              : ''
        }
        value={
          value !== undefined
            ? value !== null
              ? formatUttaksalder(intl, value, { compact: true })
              : ''
            : localValue
        }
        error={
          hasValidationError && showValidationError
            ? 'VALIDATION ERROR'
            : undefined
        }
        onChange={onLocalChange}
      >
        <option>Velg alder</option>
        {formaterteAldere.slice(0, formaterteAldere.length).map((alderChip) => (
          <option key={alderChip} value={alderChip}>
            {alderChip}
          </option>
        ))}
      </Select>
    </div>
  )
}
/* c8 ignore end */
