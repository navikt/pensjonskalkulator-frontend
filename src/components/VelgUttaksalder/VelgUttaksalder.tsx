import React from 'react'
import { FormattedMessage, useIntl } from 'react-intl'

import { Chips, Heading } from '@navikt/ds-react'
import clsx from 'clsx'

import { useAppDispatch, useAppSelector } from '@/state/hooks'
import {
  selectCurrentSimulation,
  selectNedreAldersgrense,
} from '@/state/userInput/selectors'
import { userInputActions } from '@/state/userInput/userInputSlice'
import { unformatUttaksalder } from '@/utils/alder'
import { logger } from '@/utils/logging'

import { getFormaterteAldere } from './utils'

import styles from './VelgUttaksalder.module.scss'

interface Props {
  tidligstMuligUttak?: Alder
}

export const VelgUttaksalder: React.FC<Props> = ({
  tidligstMuligUttak = { ...useAppSelector(selectNedreAldersgrense) },
}) => {
  const intl = useIntl()
  const dispatch = useAppDispatch()
  const pinRef = React.useRef<HTMLDivElement>(null)

  const { uttaksalder } = useAppSelector(selectCurrentSimulation)

  const formaterteAldere = React.useMemo(
    () =>
      getFormaterteAldere(intl, tidligstMuligUttak, { aar: 75, maaneder: 0 }),
    [tidligstMuligUttak]
  )

  const onAlderClick = (formatertAlder: string) => {
    logger('chip valgt', {
      tekst: 'Velg uttaksalder',
      data: formatertAlder,
    })
    const alder = unformatUttaksalder(formatertAlder)
    dispatch(userInputActions.setCurrentSimulationUttaksalder(alder))
    pinRef.current?.scrollIntoView({ behavior: 'smooth', block: 'start' })
  }

  return (
    <div className={styles.wrapper}>
      <div className={styles.wrapperCard}>
        <span ref={pinRef} className={styles.pin}></span>
        <Heading size="xsmall" level="2">
          <FormattedMessage id="velguttaksalder.title" />
        </Heading>
        <Chips
          className={clsx(styles.chipsWrapper, styles.chipsWrapper__hasGap)}
        >
          {formaterteAldere
            .slice(0, formaterteAldere.length)
            .map((alderChip) => {
              const { aar, maaneder } = unformatUttaksalder(alderChip)
              return (
                <Chips.Toggle
                  selected={
                    uttaksalder?.aar === aar &&
                    uttaksalder?.maaneder === maaneder
                  }
                  checkmark={false}
                  key={alderChip}
                  onClick={() => onAlderClick(alderChip)}
                >
                  {alderChip}
                </Chips.Toggle>
              )
            })}
        </Chips>
      </div>
    </div>
  )
}
