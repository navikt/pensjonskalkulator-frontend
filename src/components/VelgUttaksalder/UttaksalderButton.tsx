import React from 'react'

import { Chips } from '@navikt/ds-react'

import { formatUttaksalder } from '@/components/VelgUttaksalder/utils'

interface Props {
  alder: Uttaksalder
  isSelected: boolean
  onClick: (alder: Uttaksalder) => void
}

export const UttaksalderButton: React.FC<Props> = React.memo(
  ({ alder, isSelected, onClick }) => {
    return (
      <Chips.Toggle selected={isSelected} onClick={() => onClick(alder)}>
        {formatUttaksalder(alder)}
      </Chips.Toggle>
    )
  }
)
