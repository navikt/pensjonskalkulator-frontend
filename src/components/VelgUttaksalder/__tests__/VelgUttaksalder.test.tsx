import { describe, expect, it } from 'vitest'

import { fulfilledGetPerson } from '@/mocks/mockedRTKQueryApiCalls'
import { userInputInitialState } from '@/state/userInput/userInputSlice'
import { render, screen, userEvent } from '@/test-utils'

import { VelgUttaksalder } from '../VelgUttaksalder'

describe('VelgUttaksalder', () => {
  const uttaksalder: Alder = {
    aar: 62,
    maaneder: 10,
  }

  const mockedState = {
    api: {
      queries: {
        ...fulfilledGetPerson,
      },
    },
    userInput: {
      ...userInputInitialState,
    },
  }

  it('viser riktige aldere når uttaksalder ikke er angitt', async () => {
    render(<VelgUttaksalder tidligstMuligUttak={undefined} />, {
      // @ts-ignore
      preloadedState: {
        ...mockedState,
      },
    })
    expect(await screen.findAllByRole('button')).toHaveLength(14)
  })

  it('oppdaterer valgt knapp når brukeren velger en alder', async () => {
    const user = userEvent.setup()

    const getProps = () => ({
      tidligstMuligUttak: uttaksalder,
    })

    const { rerender } = render(<VelgUttaksalder {...getProps()} />, {
      // @ts-ignore
      preloadedState: {
        ...mockedState,
      },
    })

    await user.click(screen.getByText('65 alder.aar', { exact: false }))

    expect(window.HTMLElement.prototype.scrollIntoView).toHaveBeenCalled()

    rerender(<VelgUttaksalder {...getProps()} />)
    expect(screen.getByRole('button', { pressed: true })).toHaveTextContent(
      '65 alder.aar'
    )

    rerender(<VelgUttaksalder {...getProps()} />)
    expect(screen.getByText('72 alder.aar', { exact: false })).toBeVisible()

    await user.click(screen.getByText('72 alder.aar', { exact: false }))

    rerender(<VelgUttaksalder {...getProps()} />)
    expect(screen.getByRole('button', { pressed: true })).toHaveTextContent(
      '72 alder.aar'
    )
  })
})
