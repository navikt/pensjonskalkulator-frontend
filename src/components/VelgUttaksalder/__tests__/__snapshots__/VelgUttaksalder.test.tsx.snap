// Vitest Snapshot v1

exports[`VelgUttaksalder > Gitt at brukeren er på mobil > viser riktig label, ikon og antall knapper når brukeren ønsker å se flere aldere 1`] = `
<DocumentFragment>
  <div
    class="_wrapper_4ca3e3"
  >
    <div
      class="_wrapperCard_4ca3e3"
    >
      <span
        class="_pin_4ca3e3"
      />
      <h2
        class="navds-heading navds-heading--xsmall"
      >
        Når vil du ta ut alderspensjon?
      </h2>
      <ul
        class="navds-chips _chipsWrapper_4ca3e3 _chipsWrapper__hasGap_4ca3e3 navds-chips--medium navds-body-short navds-body-short--small"
      >
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              62 år og 10 md.
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              63 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              64 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              65 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              66 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              67 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              68 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              69 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              70 år
            </span>
          </button>
        </li>
      </ul>
      <button
        class="_visFlereAldere_4ca3e3 navds-button navds-button--tertiary navds-button--xsmall"
      >
        <span
          class="navds-button__icon"
        >
          <svg
            aria-hidden="true"
            fill="none"
            focusable="false"
            height="1em"
            role="img"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M5.97 9.47a.75.75 0 0 1 1.06 0L12 14.44l4.97-4.97a.75.75 0 1 1 1.06 1.06l-5.5 5.5a.75.75 0 0 1-1.06 0l-5.5-5.5a.75.75 0 0 1 0-1.06Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <span
          class="navds-label navds-label--small"
        >
          Vis flere aldere
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`VelgUttaksalder > Gitt at brukeren er på mobil > viser riktig label, ikon og antall knapper når brukeren ønsker å se flere aldere 2`] = `
<DocumentFragment>
  <div
    class="_wrapper_4ca3e3"
  >
    <div
      class="_wrapperCard_4ca3e3"
    >
      <span
        class="_pin_4ca3e3"
      />
      <h2
        class="navds-heading navds-heading--xsmall"
      >
        Når vil du ta ut alderspensjon?
      </h2>
      <ul
        class="navds-chips _chipsWrapper_4ca3e3 _chipsWrapper__hasGap_4ca3e3 navds-chips--medium navds-body-short navds-body-short--small"
      >
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              62 år og 10 md.
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              63 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              64 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              65 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              66 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              67 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              68 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              69 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              70 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              71 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              72 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              73 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              74 år
            </span>
          </button>
        </li>
        <li>
          <button
            aria-pressed="false"
            class="navds-chips__chip navds-chips__toggle navds-chips__toggle--action"
          >
            <span
              class="navds-chips__chip-text"
            >
              75 år
            </span>
          </button>
        </li>
      </ul>
      <button
        class="_visFlereAldere_4ca3e3 navds-button navds-button--tertiary navds-button--xsmall"
      >
        <span
          class="navds-button__icon"
        >
          <svg
            aria-hidden="true"
            fill="none"
            focusable="false"
            height="1em"
            role="img"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M11.47 7.97a.75.75 0 0 1 1.06 0l5.5 5.5a.75.75 0 1 1-1.06 1.06L12 9.56l-4.97 4.97a.75.75 0 0 1-1.06-1.06l5.5-5.5Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        <span
          class="navds-label navds-label--small"
        >
          Vis færre aldere
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
